CactbotOverlay/resources/dana/bars/bars.js

```diff
@@ -78,22 +74,131 @@ var g_init = false;
+  'Skull Sunder',
+  "Butcher's Block",
+  'Maim',
+  "Storm's Eye",
+  "Storm's Path",
+  // general (TODO: is this true?)
+  'Limit Break',
+]);
+
+class ComboTracker {
+  constructor(me, comboBreakers, callback) {
+    this.me = me;
+    this.comboTimer = null;
+    this.kReEndCombo = new RegExp(':' + me + '( starts using |:' + kReAbilityCode + ':)(' + comboBreakers.join('|') + ')( |:)');
+    this.comboNodes = {}; // { key => { re: string, next: [node keys], last: bool } }
+    this.startList = [];
+    this.callback = callback;
+    this.current = null;
+    this.considerNext = this.startList;
+  }
+
+  AddCombo(skillList) {
+    // Due to this bug: https://github.com/ravahn/FFXIV_ACT_Plugin/issues/100
+    // We can not look for log messages from FFXIV "You use X" here. Instead we
+    // look for the actual ability usage provided by the XIV plugin.
+    // Also, the networked parse info is given much quicker than the lines from the game.
+    if (this.startList.indexOf(skillList[0]) == -1) {
+      this.startList.push(skillList[0]);
+    }
+    for (var i = 0; i < skillList.length; ++i) {
+      var node = this.comboNodes[skillList[i]];
+      if (node == undefined) {
+        node = {
+          re: new RegExp(':' + this.me + ':' + kReAbilityCode + ':' + skillList[i] + ':'),
+          next: []
+        };
+        this.comboNodes[skillList[i]] = node;
+      }
+      if (i != skillList.length - 1) {
+        node.next.push(skillList[i + 1]);
+      } else {
+        node.last = true;
+      }
+    }
+  }
+
+  ParseLog(log) {
+    for (var i = 0; i < this.considerNext.length; ++i) {
+      var next = this.considerNext[i];
+      if (log.search(this.comboNodes[next].re) >= 0) {
+        this.StateTransition(next);
+        return true;
+      }
+    }
+    for (var i = 0; i < this.startList.length; ++i) {
+      var next = this.startList[i];
+      if (log.search(this.comboNodes[next].re) >= 0) {
+        this.StateTransition(next);
+        return true;
+      }
+    }
+    if (log.search(this.kReEndCombo) >= 0) {
+      this.AbortCombo();
+      return true;
+    }
+    return false;
+  }
+
+  StateTransition(nextState) {
+    if (this.current == nextState)
+      return;
+
+    window.clearTimeout(this.comboTimer);
+    this.comboTimer = null;
+    this.current = nextState;
+    this.comboTimer = null;
+
+    if (nextState == null) {
+      this.considerNext = [];
+    } else {
+      this.considerNext = this.comboNodes[nextState].next;
+      if (!this.comboNodes[nextState].last) {
+        var kComboDelayMs = 12000;
+        this.comboTimer = window.setTimeout(this.AbortCombo.bind(this), kComboDelayMs);
+      }
+    }
+    this.callback(nextState);
+  }
 
+  AbortCombo() {
+    this.StateTransition(null);
+  }
+}
+
+function setupComboTracker(me, callback) {
+  var comboTracker = new ComboTracker(me, kComboBreakers, callback);
+  comboTracker.AddCombo(['(Enchanted )?Riposte', '(Enchanted )?Zwerchhau', '(Enchanted )?Redoublement']);
```

Comment:
The RDM combo has to be Enchanted, and includes Verflare/Verholy at the end.

I didn't try to track the last step in the combo before, I treated it as breaking the combo as the outcome is the same. But if you're going to track the last bit then it should be there.

Though TBH I think in the end you're not using the .last field. So you could make these combos all skip the last one if you wanted, for warrior too. (But still require Enchanted.)

Then this made me wonder about tracking the last one with a timeout.. are your war combos 4 skills long and you omitted the last one here? If not then making a timeout on the actual last thing (like Verflare/holy for rdm) doesn't make sense to me, as the combo is already over. So maybe you should condition that on .last.

I guess there's 2 options:
1. The AddCombo omits the last thing. Make sure it's in the combo breakers set. Use timer on the last thing.
2. The AddCombo includes the last thing. Don't use a timer on it. Reset the combo to empty when it's run.

Right now I think this is a mix of the two?